sermant-plugins/sermant-mq-grayscale/sermant-mq-grayscale-plugin/src/main/java/io/sermant/mq/grayscale/interceptor/MqSchedulerRebuildSubscriptionInterceptor.java

```diff
@@ -0,0 +1,89 @@
+
+import io.sermant.core.common.LoggerFactory;
+import io.sermant.core.plugin.agent.entity.ExecuteContext;
+import io.sermant.core.plugin.agent.interceptor.AbstractInterceptor;
+import io.sermant.core.utils.StringUtils;
+import io.sermant.mq.grayscale.service.MqConsumerGroupAutoCheck;
+import io.sermant.mq.grayscale.utils.MqGrayscaleConfigUtils;
+import io.sermant.mq.grayscale.utils.SubscriptionDataUtils;
+
+import org.apache.rocketmq.client.impl.consumer.RebalanceImpl;
+import org.apache.rocketmq.common.protocol.heartbeat.SubscriptionData;
+
+import java.util.Locale;
+import java.util.concurrent.ConcurrentMap;
+import java.util.logging.Logger;
+
+/**
+ * TAG/SQL92 query message statement interceptor
+ *
+ * @author chengyouling
+ * @since 2024-05-27
+ **/
+public class MqSchedulerRebuildSubscriptionInterceptor extends AbstractInterceptor {
+    private static final Logger LOGGER = LoggerFactory.getLogger();
+
+    private final Object lock = new Object();
+
+    @Override
+    public ExecuteContext before(ExecuteContext context) throws Exception {
+        return context;
+    }
+
+    @Override
+    public ExecuteContext after(ExecuteContext context) throws Exception {
+        if (MqGrayscaleConfigUtils.isPluginEnabled()) {
+            ConcurrentMap<String, SubscriptionData> map = (ConcurrentMap<String, SubscriptionData>) context.getResult();
+            RebalanceImpl balance = (RebalanceImpl) context.getObject();
+            for (SubscriptionData subscriptionData : map.values()) {
+                if (balance.getConsumerGroup() == null) {
+                    continue;
+                }
+                if (!SubscriptionDataUtils.EXPRESSION_TYPE_SQL92.equals(subscriptionData.getExpressionType())
+                        && !SubscriptionDataUtils.EXPRESSION_TYPE_TAG.equals(subscriptionData.getExpressionType())) {
+                    LOGGER.warning(String.format(Locale.ENGLISH, "can not process expressionType: %s",
+                            subscriptionData.getExpressionType()));
+                    continue;
+                }
+                buildSql92SubscriptionData(subscriptionData, balance);
+            }
+        }
+        return context;
+    }
+
+    private void buildSql92SubscriptionData(SubscriptionData subscriptionData, RebalanceImpl balance) {
+        synchronized (lock) {
+            if (SubscriptionDataUtils.getGrayTagChangeFlag(subscriptionData.getTopic(), balance)) {
+                if (StringUtils.isEmpty(MqGrayscaleConfigUtils.getGrayGroupTag())) {
+                    MqConsumerGroupAutoCheck.setMqClientInstance(subscriptionData.getTopic(),
+                            balance.getConsumerGroup(), balance.getmQClientFactory());
+                }
+                String addrTopicGroupKey = SubscriptionDataUtils.buildAddrTopicGroupKey(subscriptionData.getTopic(),
+                        balance.getConsumerGroup(), balance.getmQClientFactory().getClientConfig().getNamesrvAddr());
```

Comment:
why you don't define multiple usage elements as variables，like  balance.getConsumerGroup()， alance.getmQClientFactory().getClientConfig().getNamesrvAddr()

Reply:
fixed